filters/auth/oidc.go

```diff
@@ -0,0 +1,546 @@
+)
+
+var (
+	src      = rand.NewSource(time.Now().UnixNano())
+	stateMap = make(map[string]bool)
+)
+
+// https://stackoverflow.com/questions/22892120/how-to-generate-a-random-string-of-a-fixed-length-in-golang
+func randString(n int) string {
+	b := make([]byte, n)
+	// A src.Int63() generates 63 random bits, enough for letterIdxMax characters!
+	for i, cache, remain := n-1, src.Int63(), letterIdxMax; i >= 0; {
+		if remain == 0 {
+			cache, remain = src.Int63(), letterIdxMax
+		}
+		if idx := int(cache & letterIdxMask); idx < len(letterBytes) {
+			b[i] = letterBytes[idx]
+			i--
+		}
+		cache >>= letterIdxBits
+		remain--
+	}
+
+	return string(b)
+}
+
+type OauthState struct {
+	Rand        string `json:"rand"`
+	Validity    int64  `json:"validity"`
+	Nonce       string `json:"none"`
+	RedirectUrl string `json:"redirectUrl"`
+}
+
+func createState(nonce []byte, redirectUrl string) ([]byte, error) {
+	state := &OauthState{
+		Rand:        randString(secretSize),
+		Validity:    time.Now().Add(stateValidity).Unix(),
+		Nonce:       fmt.Sprintf("%x", nonce),
+		RedirectUrl: redirectUrl,
+	}
+	return json.Marshal(state)
+}
+
+func extractState(encState []byte) (*OauthState, error) {
+	var state OauthState
+	err := json.Unmarshal(encState, &state)
+	if err != nil {
+		return nil, err
+	}
+	return &state, nil
+}
+
+func (f *tokenOidcFilter) internalServerError(ctx filters.FilterContext) {
+	rsp := &http.Response{
+		StatusCode: http.StatusInternalServerError,
+	}
+	ctx.Serve(rsp)
+}
+
+func (f *tokenOidcFilter) doOauthRedirect(ctx filters.FilterContext) {
+	nonce, err := f.encrypter.createNonce()
+	if err != nil {
+		log.Errorf("Failed to create nonce: %v", err)
+		f.internalServerError(ctx)
+		return
+	}
+
+	redirectUrl := ctx.Request().URL.String()
+	statePlain, err := createState(nonce, redirectUrl)
+	if err != nil {
+		log.Errorf("failed to create oauth2 state: %v", err)
+		f.internalServerError(ctx)
+		return
+	}
+	stateEnc, err := f.encrypter.encryptDataBlock(statePlain)
+	if err != nil {
+		log.Errorf("Failed to encrypt data block: %v", err)
+		f.internalServerError(ctx)
+		return
+	}
+
+	oauth2URL := f.config.AuthCodeURL(fmt.Sprintf("%x", stateEnc), f.authCodeOptions...)
+	rsp := &http.Response{
+		Header: http.Header{
+			"Location": []string{oauth2URL},
+		},
+		StatusCode: http.StatusTemporaryRedirect,
+		Status:     "Moved Temporarily",
+	}
+	log.Infof("serve redirect: plaintextState:%s to Location: %s", statePlain, rsp.Header.Get("Location"))
+	ctx.Serve(rsp)
+}
+
+func (f *tokenOidcFilter) Response(filters.FilterContext) {}
+
+func (f *tokenOidcFilter) doDownstreamRedirect(ctx filters.FilterContext, oidcState []byte, redirectUrl string) {
+	log.Debugf("Doing Downstream Redirect to :%s", redirectUrl)
+	r := &http.Response{
+		StatusCode: http.StatusTemporaryRedirect,
+		Header: map[string][]string{
+			"Set-Cookie": {fmt.Sprintf("%s=%x; Path=/; HttpOnly; MaxAge=%d", f.cookiename, oidcState, int(f.validity.Seconds()))},
```

Comment:
Do I understand correctly that by changing the validity seconds, a user can manipulate the session lifetime?